### PR TITLE
RAD-113 Support images as concepts

### PR DESCRIPTION
**Pull request by bgeVam**
see https://issues.openmrs.org/browse/RAD-113

### CODE REVIEW

File: omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java [hunk 1]
```diff
@@ -351,30 +349,28 @@ public void saveObs_ShouldEditObsWithEditReasonComplexConceptANdRequestWhichIsAn
...
 		when(obsErrors.hasErrors()).thenReturn(false);
-		ConceptComplex concept = new ConceptComplex();
-		ConceptDatatype cdt = new ConceptDatatype();
-		cdt.setHl7Abbreviation("ED");
-		concept.setDatatype(cdt);
-		mockObs.setConcept(concept);
+		
 		final String fileName = "test.txt";
 		final byte[] content = "Hello World".getBytes();
 		MockMultipartFile mockMultipartFile = new MockMultipartFile("complexDataFile", fileName, "text/plain", content);
```
Comment (teleivo) [inline omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java]: @bgeVam remove empty line for readability (closeness of mockMultipartFile, mockRequest.addFile(mockMultipartFile))

File: omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java [hunk 2]
```diff
@@ -351,30 +349,28 @@ public void saveObs_ShouldEditObsWithEditReasonComplexConceptANdRequestWhichIsAn
 		mockRequest.setSession(mockSession);
 		
 		when(obsErrors.hasErrors()).thenReturn(false);
```
Comment (teleivo) [inline omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java]: @bgeVam move when(..) down, when's should be grouped together

File: omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java [hunk 3]
```diff
@@ -433,10 +429,42 @@ public void saveObs_ShouldPopulateModelAndViewWithObsOccuringThrownAPIException(
...
+	 * @verifies return populated model and view if binding errors occur for complex concept
+	 */
+	@Test
+	public void saveObs_shouldReturnPopulatedModelAndViewIfBindingErrorsOccurForComplexConcept() {
+		
+		MockHttpSession mockSession = new MockHttpSession();
+		MockMultipartHttpServletRequest mockRequest = new MockMultipartHttpServletRequest();
+		mockRequest.addParameter("saveObs", "saveObs");
+		mockRequest.setSession(mockSession);
+		final String fileName = "test.txt";
```
Comment (teleivo) [inline omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java]: add newline

File: omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java [hunk 4]
```diff
@@ -433,10 +429,42 @@ public void saveObs_ShouldPopulateModelAndViewWithObsOccuringThrownAPIException(
...
+		final byte[] content = "Hello World".getBytes();
+		MockMultipartFile mockMultipartFile = new MockMultipartFile("complexDataFile", fileName, "text/plain", content);
+
```
Comment (teleivo) [inline omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java]: @bgeVam remove empty line for readability (closeness of mockMultipartFile, mockRequest.addFile(mockMultipartFile))

File: omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java [hunk 5]
```diff
@@ -457,22 +485,26 @@ public void populateModelAndView_ShouldPopulateModelAndViewWithObsForGivenRadiol
...
 		assertThat(modelAndView.getModelMap(), hasKey("previousObs"));
 		List<Obs> previousObs = (List<Obs>) modelAndView.getModelMap().get("previousObs");
-		assertNotNull(previousObs);
+		assertThat(previousObs, is(notNullValue()));
 		assertThat(previousObs, is(radiologyService.getObsByOrderId(mockRadiologyOrder.getId())));
 		
 		assertThat(modelAndView.getModelMap(), hasKey("dicomViewerUrl"));
+		
+		assertThat(modelAndView.getModelMap(), hasKey(not("htmlView")));
+
```
Comment (teleivo) [inline omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java]: remove newlines

File: omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java [hunk 6]
```diff
@@ -495,17 +527,64 @@ public void populateModelAndView_ShouldPopulateModelAndViewWithObsForGivenRadiol
...
 		List<Obs> previousObs = (List<Obs>) modelAndView.getModelMap().get("previousObs");
-		assertNotNull(previousObs);
+		assertThat(previousObs, is(notNullValue()));
 		assertThat(previousObs, is(radiologyService.getObsByOrderId(mockRadiologyOrder.getId())));
 		
 		assertThat(modelAndView.getModelMap(), hasKey("dicomViewerUrl"));
+		
+		assertThat(modelAndView.getModelMap(), hasKey(not("htmlView")));
+		
+		assertThat(modelAndView.getModelMap(), hasKey(not("hyperlinkView")));
```
Comment (teleivo) [inline omod/src/test/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormControllerTest.java]: remove newlines between the 3 assertThats

File: omod/src/main/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormController.java [hunk 7]
```diff
@@ -102,6 +103,17 @@ private ModelAndView populateModelAndView(RadiologyOrder radiologyOrder, Obs obs
 		result.addObject("studyUID", study.isCompleted() ? study.getStudyInstanceUid() : null);
 		result.addObject("dicomViewerUrl", getDicomViewerUrl(study, radiologyOrder.getPatient()));
 		
+		if (obs.getId() != null && obs.getConcept().isComplex()) {
+			Obs complexObsAsHtmlView = obsService.getComplexObs(Integer.valueOf(obs.getId()), WebConstants.HTML_VIEW);
+			ComplexData complexData = complexObsAsHtmlView.getComplexData();
+			result.addObject("htmlView", complexData.getData());
```
Comment (teleivo) [inline omod/src/main/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormController.java]: @bgeVam please inline complexObsAsHtmlView.getComplexData().getData()

File: omod/src/main/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormController.java [hunk 8]
```diff
@@ -102,6 +103,17 @@ private ModelAndView populateModelAndView(RadiologyOrder radiologyOrder, Obs obs
...
+			
+			Obs complexObsAsHyperlinkView = obsService.getComplexObs(Integer.valueOf(obs.getId()),
+			    WebConstants.HYPERLINK_VIEW);
+			ComplexData complexData2 = complexObsAsHyperlinkView.getComplexData();
+			result.addObject("hyperlinkView", complexData2.getData());
```
Comment (teleivo) [inline omod/src/main/java/org/openmrs/module/radiology/web/controller/RadiologyObsFormController.java]: same as above

**Comment by bgeVam** [comment]
PR Updated!

**Pull request merged**